File: src/main/java/org/example/programmers/다리를지나는트럭.java
package org.example.programmers;

import java.util.*;

public class 다리를지나는트럭 {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;

        Queue<Integer> truck_queue = new LinkedList<>();
        for (int i : truck_weights) {
            truck_queue.add(i);
        }



        Queue<Integer> bridge_queue = new LinkedList<>();

        for (int i = 0; i < bridge_length; i++) {
            bridge_queue.add(0);
        }
        int current_weight = 0;

        while (!truck_queue.isEmpty()) {
            int done = bridge_queue.poll();
            answer++;
            current_weight -= done;
        }

        return answer;


    }

    public int solution2(int bridge_length, int weight, int[] truck_weights) {
        int time = 0;
        int currentWeight = 0;

        Queue<Integer> waitingTrucks = new LinkedList<>(); // 대기 중인 트럭
        Queue<Integer> bridgeTrucks = new LinkedList<>(); // 다리 위의 트럭
        Queue<Integer> timeOnBridge = new LinkedList<>(); // 각 트럭이 다리에 올라간 시간

        // 대기열에 트럭 추가
        for (int truckWeight : truck_weights) {
            waitingTrucks.add(truckWeight);
        }

        // 다리는 초기에 비어있음 (0으로 초기화)
        for (int i = 0; i < bridge_length; i++) {
            bridgeTrucks.add(0);
        }

        // 모든 트럭이 다리를 건널 때까지 반복
        while (!bridgeTrucks.isEmpty()) {
            time++; // 시간 증가

            // 다리의 맨 앞 트럭이 나감
            currentWeight -= bridgeTrucks.poll();

            // 대기 중인 트럭이 있고 다리에 더 올릴 수 있는 경우
            if (!waitingTrucks.isEmpty() && currentWeight + waitingTrucks.peek() <= weight) {
                int newTruck = waitingTrucks.poll();
                bridgeTrucks.add(newTruck);
                currentWeight += newTruck;
            } else {
                // 새 트럭을 올릴 수 없으면 0 추가 (빈 공간)
                bridgeTrucks.add(0);
            }

            // 모든 대기 트럭이 다리에 올라갔고, 다리 위에 트럭이 없으면 종료
            if (waitingTrucks.isEmpty() && currentWeight == 0) {
                break;
            }
        }

        return time;
    }

    public static void main(String[] args) {
        다리를지나는트럭 s = new 다리를지나는트럭();
        System.out.println(s.solution(2, 10, new int[]{7, 4, 5, 6}));
        System.out.println(s.solution(100, 100, new int[]{10}));
        System.out.println(s.solution(100, 100, new int[]{10, 10, 10, 10, 10, 10, 10, 10, 10, 10}));
    }

}

File: src/main/python/programmers/다리를지나는트럭.py
"""
다리를 지나는 트럭
프로그래머스 Lv2
https://school.programmers.co.kr/learn/courses/30/lessons/42583?language=python3
"""

from collections import deque


def solution(bridge_length, weight, truck_weights):
    """
    거참까다롭네.
    """

    answer = 0
    bridge_queue = deque()
    # 초기 상태.
    for _ in range(bridge_length):
        bridge_queue.append(0)

    truck_queue = deque()
    for truck in truck_weights:
        truck_queue.append(truck)
    time = 0
    current_weight = 0

    while truck_queue:
        # bridge_stack 에서 leftpop
        current_weight -= bridge_queue.popleft()
        # time ++
        time += 1
        # truck_weight[0] 에다가 현재 다리 무게를 더해서 weight를 안넘으면 bridge_weight에 추가 (weight 를 안넘을 때만)
        if current_weight + truck_queue[0] <= weight:
            # 안넘으면 bridge_queue에도 추가.
            # truck_weights에서 leftpop
            truck = truck_queue.popleft()
            bridge_queue.append(truck)
            current_weight += truck

        # 넘으면 bridge_queue에 0 추가
        else:
            bridge_queue.append(0)

        #

    # truck_weights가 빈 순간, 즉 마지막 팝레프트 된순간 종료되므로, 마지막 애가 지나가는 시간만큼 더해줘야함
    answer = time + bridge_length
    return answer


print(solution(2, 10, [7, 4, 5, 6]))
print(solution(100, 100, [10]))
print(solution(100, 100, [10, 10, 10, 10, 10, 10, 10, 10, 10, 10]))
